File: core/java/src/main/java/hara/lang/protocol/IAbort.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Exceptional stream termination capability. */
@HaraProtocolBinding(
    namespace = "std.protocol.iabort",
    name = "IAbort",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IAbort {
  @HaraMethod(value = "abort", arity = 2)
  Object abort(Object error);
}

File: core/rust/src/lang/protocol/iabort.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iabort",
    name = "IAbort",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IAbort {
    type Error;
    type Output;

    #[hara_method(value = "abort", arity = 2)]
    fn abort(&self, error: Self::Error) -> Self::Output;
}
